### ISSUE TITLE
workers located at same host converge better than workers at different hosts

### ISSUE THREAD

**Issue by glkclass** [uber/horovod#173]
Hi,
running benchmarks from https://github.com/alsrgv/benchmarks/tree/horovod_v2 using horovod approach I see that configurations with workers located at same host converge better comparing to same number of workers but located at different hosts.
Hence the question: Do are math operations (gradient averaging, ...) the fully equivalent when workers are located on same host and on different hosts?

![figure_1](https://user-images.githubusercontent.com/5112030/36147181-e75c2d66-10bf-11e8-9c9c-da2cec35b263.png)

7578829_btch_glb128_w4xb32 (green plot, 4 workers are at same host) and 7578686_btch_glb128_w4xb32 (cyan plot, 4 workers occupy two hosts(2x2 config) ) have same number of workes with same input batch but different convergence curve. 
And this can be observed for other cases. 

I am running resnet101 benchmark training with following command lines:
Case 1:
`mpirun  -np 2 -host host0 -env CUDA_VISIBLE_DEVICES=2,3 python tf_cnn_benchmarks.py --flush_stdout --model resnet101 --data_dir dataset/imagenet --batch_size 32 --num_batches 500000 --variable_update horovod --horovod_device cpu 
: -np 2 -host host1 -env CUDA_VISIBLE_DEVICES=1,3 python tf_cnn_benchmarks.py --flush_stdout --model resnet101 --data_dir dataset/imagenet --batch_size 32 --num_batches 500000 --variable_update horovod --horovod_device cpu`

Case 2:
`mpirun  -np 4 -host host2 -env CUDA_VISIBLE_DEVICES=0,1,2,3 python tf_cnn_benchmarks.py --flush_stdout --model resnet101 --data_dir dataset/imagenet --batch_size 32 --num_batches 500000 --variable_update horovod --horovod_device cpu`

Thanks.

**Comment by alsrgv** [comment]
Math operations are supposed to be the same, so this is very interesting.  Do you observe the same behavior with NCCL?

**Comment by glkclass** [comment]
@alsrgv  I have Intel MPI, so installed Horovod using next command:
`HOROVOD_GPU_ALLREDUCE=MPI pip install horovod`
That means I don't use NCCL for all-reduce op  and I can't check it right now.
As I understood NCCL/NCCL2 is required only to perform all-reduce op fast 
(please, correct me if I am wrong)

Do you guess that different all-reduce implementations may lead to such significant difference ?

**Comment by alsrgv** [comment]
@glkclass, a couple of things to try and see if you observe any difference:
1. Use Intel MPI on CPU, i.e. install with: `pip install --no-cache-dir horovod`.  I'm curious if there is any discrepancy between CPU and GPU - MPI implementations often have bugs in GPU implementations.
2. Try NCCL2 - you should be able to do that irrespective of what MPI you're using.

**Comment by glkclass** [comment]
@alsrgv , 
1. Do you mean to set flag horovod_allreduce=CPU ?
2. How can I force application to use nccl2 allrefuce instead Intel MPI allreduce? Any flag?

**Comment by alsrgv** [comment]
1. You can set `--horovod_device cpu` in benchmarks.
2. Uninstall the current Horovod using `pip uninstall -y horovod` and then follow [this guide](https://github.com/uber/horovod/blob/master/docs/gpus.md#have-gpus) to setup Horovod with NCCL2.  You don't need to install Open MPI - this should work for Intel MPI as well.

**Comment by glkclass** [comment]
1. --horovod_device = cpu/gpu - seems doesn't matter.
Have run 4 workers on 1/2/4 hosts with different `--horovod_device` flag. Results are attached. There is obvious difference between them (red, green, blue plots) irrespective to  `--horovod_device` flag.
 
2. Will try to do this later and publish results here.

![figure_1](https://user-images.githubusercontent.com/5112030/36372246-22bf3b3a-156d-11e8-941b-0d5c3020b60c.png)

**Comment by alsrgv** [comment]
Is total # of GPUs the same if you run on 1/2/4 hosts?  If not, you should adjust learning rate to be proportional to the # of GPUs.

E.g. if you learning rate on 1 GPU with batch size 32 is 0.01, your learning rate on 4 GPUs with batch size 32 should be 0.04.

**Comment by glkclass** [comment]
@alsrgv, Total number of GPUs(workers) is 4 for all cases. They just located on 1/2/4 hosts: i.e. 4x1, 2x2, 1x4. From math point of view, they should be equivalent.

**Comment by alsrgv** [comment]
@glkclass, thanks for the confirmation.  Please let me know how NCCL test goes.  I will run similar tests on my side.

**Comment by glkclass** [comment]
@alsrgv , Ok. Will update you with nccl results, but maybe in several days.

**Comment by alsrgv** [comment]
@glkclass, I'm running a repro using https://github.com/tensorflow/benchmarks.

I modified a couple of things: 1) there's no `--flush_stdout` anymore, 2) I'm using fewer batches, 3) my dataset is 10% of ImageNet, 4) I'll use TF summaries and TensorBoard, 5) I'm using fixed learning rate.

One node:
```
$ mpirun -np 4 python tf_cnn_benchmarks.py \
  --model resnet101 \
  --data_dir ~/train_data \
  --batch_size 32 \
  --num_batches 100000 \
  --variable_update horovod \
  --horovod_device cpu \
  --data_name imagenet \
  --summary_verbosity 1 \
  --save_summaries_steps 100 \
  --train_dir single_node \
  --learning_rate 0.05
```

Four nodes:
```
$ mpirun -np 4 -H node1:1,node2:1,node3:1,node4:1 python tf_cnn_benchmarks.py \
  --model resnet101 \
  --data_dir ~/train_data \
  --batch_size 32 \
  --num_batches 100000 \
  --variable_update horovod \
  --horovod_device cpu \
  --data_name imagenet \
  --summary_verbosity 1 \
  --save_summaries_steps 100 \
  --train_dir single_node \
  --learning_rate 0.05
```

Will update you when these finish.

**Comment by alsrgv** [comment]
My results are opposite to yours - on multiple nodes I got better loss than on a single node.

<img width="1038" alt="screen shot 2018-02-26 at 12 42 59 am" src="https://user-images.githubusercontent.com/16640218/36661266-af83ce06-1a8f-11e8-808c-0ab7923c7558.png">

I will do a pair of runs with NCCL to see if there will be any difference in the behavior.

**Comment by glkclass** [comment]
@alsrgv, In fact our results are the same: they show difference in learning process for two cases which supposed to be equivalent from this point of view ). 
Unfortunately I wasn't able to try with NCCL runs due to some issues with our infrastructure. Will do it when have a chance and upload results here.

**Comment by alsrgv** [comment]
I think I found the root cause of the problem.  

```python
    tf.set_random_seed(self.params.tf_random_seed)
    np.random.seed(4321)
```

It appears that the way TF/numpy seed works, random generator is not guaranteed to generate the same numbers on different machines.  So, if we run four processes on the same machine, we get all of them to read the **same** TF records.  However, if we run on different machines, they read different TF records in the same batch.

I have modified seed to always be unique and depend on Horovod rank, and now training is comparable in both configurations. Also, it matches "multi-host" curve from before.

```python
    import horovod.tensorflow as hvd
    tf.set_random_seed(self.params.tf_random_seed + hvd.rank())
    np.random.seed(4321 + hvd.rank())
```

<img width="1035" alt="screen shot 2018-02-28 at 12 03 52 am" src="https://user-images.githubusercontent.com/16640218/36776562-1b45cd06-1c1b-11e8-90b8-794fdbfc2789.png">

I plan on sending TF benchmarks a PR with this fix when I get a chance.

**Comment by glkclass** [comment]
@alsrgv , Will try your fix at my side and update you with results.

Meanwhile, want to clarify some moments here...
1. As I understand all workers initialize the network using same set of parameters (weights). i.e. there is a main worker which initializes its network with set of random values and broadcasts this initial set of parameters to every worker inside the horovod. I.e. every worker has the same initial model. So there shouldn't be any issues with different random values across the nodes. 
Please, correct me if I am wrong.

2. If I understood your guess correctly, you mean than all the workers running on the same node use same set of input images as input mini-batch due to same random value generator initial stage. And workers running on different nodes use different set of images as mini-batch since the random generator initialized differently on different nodes . And this is a reason of training curve discrepancy.

But this mean that 
a) during the training on **different** nodes every worker reads **different** mini-batch, based on this mini-batch computes **different** gradient updates to be averaged after.
And this is really distributed training.

b) during the training on single node every worker reads **same** mini-batch. Every worker processes **same** set of input images and produces **same** gradient updates to be averaged after. So we don't benefit from multi-worker computation since every worker computes same gradient updates.
And this is equivalent to single worker training.

My understanding that every worker should process different mini-batch and produce different gradient updates which will be averaged after. Thus leading to more precise gradient updates due to larger effective input mini-batch.

Please, correct me if I am wrong.

**Comment by alsrgv** [comment]
Your clarifications are exactly correct.  They're further confirmed by the fact that loss curve with seed adjusted to rank matches loss curve of original TF benchmark executed on multiple nodes.

**Comment by glkclass** [comment]
But this seems to contradict my results. On the other hand it may be an issue of our supercom infrastructure: jobs scheduled for execution  to different nodes have same seed, while jobs on same node have different seeds. At any case, adding rank makes seed unique and should fix this problem. Will run tests and update this thread when ready. Thanks.

**Comment by glkclass** [comment]
Sorry for late updating the discussion. 

Have the following results:
1. Running tests using **openMPI+NCCL** ( i.e. Horovod installation with `HOROVOD_GPU_ALLREDUCE=NCCL` which should mean: perform all-reduce operation using NCCL lib)  **have no this issue** (difference between single and multiple host runs). Even **without hvd.rank mixing together with rnd seed**

2. Running tests using **IntelMPI** ( i.e. Horovod installation with `HOROVOD_GPU_ALLREDUCE=MPI` which should mean: perform all-reduce operation using IntelMPI lib)  continue having the issue **even after hvd.rank mixing together with rnd seed**

3. **--horovod_device cpu/gpu** doesn't matter in both cases.

4. Seems runs using **IntelMPI** have better convergence comparing to **openMPI+NCCL** per same number of steps using same batch size

Will continue experiments when nave a chance.

**Comment by alsrgv** [comment]
Update: the random seed fix has been merged into TF benchmarks.

@glkclass, I'm not sure I completely follow.  If I take (1) + (3), it means that Open MPI with and without NCCL has no issue.  However, you mention that Intel MPI does have an issue in (2).  Finally, in (4) you mention that in fact, Intel MPI converges better than Open MPI.  Can you clarify your experiments?

**Comment by glkclass** [comment]
@alsrgv 
1. I used your appropriate/temporary branch 'different_seed_for_different_workers' for experiments

2. Runs using Horovod version installed as following: `HOROVOD_GPU_ALLREDUCE=NCCL pip install --no-cache-dir horovod` **have no this issue**. I suppose all-reduce operation is executed by NCCL2 lib here.

3. Runs using Horovod version installed as following: `HOROVOD_GPU_ALLREDUCE=MPI HOROVOD_GPU_ALLGATHER=MPI HOROVOD_GPU_BROADCAST=MPI pip install --no-cache-dir horovod` **have his issue and I couldn't fix** it by mixing hvd.rank with seed. I suppose all-reduce operation is executed by our proprietary MPI (IntelMPI) lib here.

4. Comparing **equivalent** (same batch, same worker/hosts config) train curves for these two Horovod installs I see that training with all-reduce operation performed on IntelMPI lib converge faster. It doesn't related to playing with different worker/host combinations. Just observation when looking on plots.

**Comment by alsrgv** [comment]
In (3) you're saying that Intel MPI with GPU has issue, but in (4) you say it converges faster than Open MPI.  Do you mean Intel MPI with NCCL?

We're you doing git pulls from my fork of benchmarks between your tests? As part of addressing code review comments there was a time when the code had regressed to an initial state.  I wonder if it may have had an impact on your tests.

**Comment by alsrgv** [comment]
I see, that's very interesting! Can you share the curves for Open MPI + NCCL and Intel MPI + GPU?

**Comment by glkclass** [comment]
@alsrgv, I will rerun tests one more again to be sure. Could point me at last relevant version of benchmarks on github?

**Comment by alsrgv** [comment]
@glkclass, `master` from https://github.com/tensorflow/benchmarks should be working fine now.

**Comment by glkclass** [comment]
@alsrgv , seems I can't reproduce the difference between **OpenMPI+NCCL vs IntelMPI**.
Let's consider I was wrong  here and we have no diff between runs on different MPI

**Comment by glkclass** [comment]
@alsrgv , Yes I can't reproduce the difference between runs on IntelMPI vs OpenMPI_NCCL. So let's consider we have no this issue for now.
